lib/api/axiosFactory.js

```diff
@@ -0,0 +1,274 @@
+    axiosInstance = this.#addPartnerTokenRefresher(axiosInstance);
+
+    return axiosInstance;
+  }
+
+  // Add a response interceptor to refresh the partner token if it's expired
+  static #addPartnerTokenRefresher(axiosInstance) {
+    axiosInstance.interceptors.response.use(
+      (response) => {
+        return response;
+      },
+      async (error) => {
+        const originalConfig = error.config;
+        if (error.response) {
+          if (error.response.status === 401 && !originalConfig._retry) {
+            // Set _retry to true after trying the first refresh request on 401 status to avoid an infinite loop
+            originalConfig._retry = true;
+
+            try {
+              // Get a refresh API key
+              const originalPartnerId = originalConfig.params.partner_id;
+              const originalPartnerApiKey = originalConfig.params.api_key;
+
+              const response = await axiosInstance.post(
+                `${this.BASE_URL}/api/partner/v1/refresh_api_key?api_key=${originalPartnerApiKey}`
+              );
+
+              firmCredentials.storePartnerApiKey(
+                originalPartnerId,
+                response.data.api_key
+              );
+
+              consola.debug("Refreshed partner api key");
+
+              // Set the new access token for current and future requests
+              axiosInstance.defaults.params.api_key = response.data.api_key;
+              originalConfig.params.api_key = response.data.api_key;
+
+              return axiosInstance(originalConfig);
+            } catch (_error) {
+              consola.error(
+                `Error 401: Failed to refresh the partner API key automatically, try to manually authorize the partner again with the authorize-partner command`
+              );
+
+              if (_error.response && _error.response.data) {
+                return Promise.reject(_error.response.data);
+              }
+
+              return Promise.reject(_error);
+            }
+          }
+        }
+
+        return Promise.reject(error);
+      }
+    );
+
+    return axiosInstance;
+  }
+
+  /**
+   * Check if current host is staging or not
+   * @returns {Boolean}
+   */
+  static #isStaging() {
+    return /staging\.getsilverfin/.test(this.BASE_URL);
+  }
+
+  static #basicAuthHeader() {
+    if (!process.env.SF_BASIC_AUTH) {
+      consola.error(`Missing environment variable: SF_BASIC_AUTH`);
+      process.exit(1);
+    }
+    return `Basic ${process.env.SF_BASIC_AUTH}`;
+  }
+
+  /**
+   * Get a new set of tokens for a given firm id
+   */
+  static async #refreshFirmTokens(axiosInstance, firmId) {
+    try {
+      consola.debug(`Refreshing tokens for firm ${firmId}`);
+      const firmTokens = firmCredentials.getTokenPair(firmId);
+      let data = {
+        client_id: process.env.SF_API_CLIENT_ID,
+        client_secret: process.env.SF_API_SECRET,
+        redirect_uri: "urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob",
```

Comment:
Is this url encoded? Can we add the raw path instead and then run something like `.urlEncode()` (not sure this exists but you get the idea) for better readability?

Reply:
Updated